src/org/rust/lang/icons/RustIconProvider.kt

```diff
@@ -1,17 +1,60 @@
 package org.rust.lang.icons
 
 import com.intellij.ide.IconProvider
+import com.intellij.openapi.util.Iconable
 import com.intellij.psi.PsiElement
-import org.rust.lang.core.psi.impl.RustFnItemImpl
-import org.rust.lang.core.psi.impl.RustImplMethodImpl
+import com.intellij.psi.util.PsiUtil
+import com.intellij.ui.RowIcon
+import com.intellij.util.PlatformIcons
+import com.intellij.util.VisibilityIcons
+import com.intellij.util.ui.EmptyIcon
+import org.rust.lang.core.psi.*
 import javax.swing.Icon
 
 class RustIconProvider: IconProvider() {
     override fun getIcon(element: PsiElement, flags: Int): Icon? {
         return when (element) {
-            is RustFnItemImpl -> RustIcons.FUNCTION
-            is RustImplMethodImpl -> RustIcons.METHOD
+            is RustEnumItem -> getEnumIcon(element, flags)
+            is RustEnumDef -> RustIcons.FIELD
+            is RustStructItem -> getStructIcon(element, flags)
+            is RustStructDeclField -> getStructDeclFieldIcon(element, flags)
+            is RustFnItem -> RustIcons.FUNCTION
+            is RustImplMethod -> RustIcons.METHOD
             else -> null
         }
     }
-}
\ No newline at end of file
+
+    private fun getEnumIcon(element: RustEnumItem, flags: Int): Icon? {
+        if ((flags and Iconable.ICON_FLAG_VISIBILITY) == 0)
+            return RustIcons.ENUM;
+
+        val parent = element.parent?.parent as RustItem?
```

Comment:
I agree, it's not particularly pretty right now... How would I implement any methods on the PSI elements? I thought they were generated and not really editable.

Side question: What does PSI actually stand for?